### PR TITLE
Use mask to determine workspace gaps assignments

### LINKED ISSUE #5282: `workspace X gaps outer|inner 0` does not work
## I'm submitting a…

<pre>
[x] Bug
[ ] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

## Reproduction Instructions

```
gaps outer 20
workspace 2 gaps outer 0
workspace 2 gaps outer 1
```

1. Apply config
2. Switch to workspace 2, outer gaps are 20
3. Switch to workspace 3, outer gaps are 1

## Environment

i3 4.21

### PR DESCRIPTION
**Pull request by orestisfl**
Fixes #5282

### CODE REVIEW

File: src/config_directives.c [hunk 1]
```diff
@@ -234,67 +234,53 @@ CFGFUN(for_window, const char *command) {
...
-            } else if (!strcmp(scope, "bottom")) {
-                assignment->gaps.bottom = gaps.bottom;
-            } else if (!strcmp(scope, "left")) {
-                assignment->gaps.left = gaps.left;
-            } else {
-                ELOG("Invalid command, cannot process scope %s", scope);
-            }
-
-            return;
+            break;
```
Comment (stapelberg) [inline src/config_directives.c:267]: Here and in src/gaps.c, relying on the `assignment` loop variable and just `break`ing when finding a matching assignment looks incomplete/unsafe to me — wouldn’t the last checked (not necessarily matching!) assignment always be stored in `assignment` after the loop is over?

Can you instead use a separate variable to which matching assignments are explicitly assigned? You could also introduce a helper function for finding a workspace assignment by name.

**Review by stapelberg — COMMENTED** [review]
Thanks, this looks good overall!

File: src/config_directives.c (hunk 1 above)
Comment (orestisfl) [reply src/config_directives.c:267]: Oops, that was silly

File: src/gaps.c [hunk 2]
```diff
@@ -117,29 +117,35 @@ bool gaps_has_adjacent_container(Con *con, direction_t direction) {
...
+    bool found = false;
     struct Workspace_Assignment *assignment;
     TAILQ_FOREACH (assignment, &ws_assignments, ws_assignments) {
         if (strcmp(assignment->name, ws->name) == 0) {
             gaps = assignment->gaps;
+            found = true;
             break;
         } else if (ws->num != -1 && name_is_digits(assignment->name) && ws_name_to_number(assignment->name) == ws->num) {
             gaps = assignment->gaps;
+            found = true;
```
Comment (stapelberg) [inline src/gaps.c]: Should there also be a `break` here?

Comment (orestisfl) [reply src/gaps.c]: No, this was intentional. Exact named assignments take precedence over numerical assignments.

**Comment by orestisfl** [comment]
Further simplified the configuration part

File: src/gaps.c (hunk 2 above)
Comment (orestisfl) [reply src/gaps.c]: But there was a mistake here because the code after used `assignment` to get the mask

**Review by stapelberg — APPROVED** [review]

**Comment by stapelberg** [comment]
Thanks!

**Pull request merged**